Integrate Open-source Models

MorphlingEd opened the pull request:
## Description

- Tried to introduce open-source models into CAMEL framework. Have successfully used LLaMA2 (7b, chat) and Vicuna (7b, v1.5) for assistant role, with corresponding `ModelType` elements defined.
- The basic workflow is based on an external server running LLM inference. This is because: 
  1. such a server can keep the LLM in memory eliminating the need to repeatedly loading it from file for every request; 
  2. such type of servers typically implemented OpenAI compatible interfaces, which makes the interactions between them quite convenient. Current implementation assumes the user is running a server by himself and should provide CAMEL with the URL to this server, e.g. `http://localhost:8000/v1`
  3. This can provided some degree of freedom to choose their preferred server and decouples the inference logic of LLMs from CAMEL. My current choice is [FastChat](https://github.com/lm-sys/FastChat), which supports OpenAI compatible interfaces and many mainstream open-source models.
- Necessary arguments and how they are specified and passed:
  - path to model tokenizers: added an extra argument `model_path` for `BaseModelBackend` class, though this should only be applicable for open-source models. It can either be a model id in HuggingFace such as `lmsys/vicuna-13b-v1.5-16k` or a path to the local folder containing the tokenizer (necessary) and models.
  - URL to the server running the LLM: It should be specified in the environmental variable `OPENAI_API_BASE`, which will be set to alter `openai.api_base`. 
- Refactored the token counting functionality as different types of open-source models have different tokenization regulations. This PR  packed the related functionalities into `utils/token_counting.py` and equips each model object with a specific token counter. 
- Removed `BaseMessage.token_len` as this function seems not useful. Counting number of tokens cannot be decoupled from the specific model and especially the open-source models need to load specific tokenizers. Each time of counting tokens here needs to load the tokenizer, which is quite costly. We can just count the tokens after models are defined with their token counters.

The functionality structure may need to be further optimized but Guohao told me to create this PR for more convenient reviewing.

## Motivation and Context

Why is this change required? What problem does it solve?
close #225
- [x] I have raised an issue to propose this change ([required](https://github.com/camel-ai/camel/issues) for new features and bug fixes)

## Types of changes

What types of changes does your code introduce? Put an `x` in all the boxes that apply:

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds core functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)
- [x] Documentation (update in the documentation)
- [x] Example (update in the folder of example)

## Implemented Tasks

- [ ] Subtask 1
- [ ] Subtask 2
- [ ] Subtask 3

## Checklist

Go over all the following points, and put an `x` in all the boxes that apply.
If you are unsure about any of these, don't hesitate to ask. We are here to help!

- [x] I have read the [CONTRIBUTION](https://github.com/camel-ai/camel/blob/master/CONTRIBUTING.md) guide. (**required**)
- [x] My change requires a change to the documentation.
- [x] I have updated the tests accordingly. (*required for a bug fix or a new feature*)
- [ ] I have updated the documentation accordingly.

dandansamax commented on examples/test/test_ai_society_example.py:
```diff
@@ -26,3 +28,15 @@ def test_ai_society_role_playing_example():
         examples.function_call.role_playing_with_function.main(ModelType.STUB)
+
+
+def test_role_playing_with_llama2():
```
These two tests become meaningless with `ModelType.STUB`.

MorphlingEd replied:
Exactly... But you know I have not yet implemented the functionality to automatically launching server running LLMs as you mentioned. The tests can certainly not be passed.

dandansamax commented on camel/models/model_factory.py:
```diff
@@ -25,8 +30,11 @@ class ModelFactory:
+    def create(
+        model_type: ModelType,
+        model_config_dict: Dict,
+        model_path: Optional[str] = None,
```
Is it possible to not contain `model_path` parameter in this? I think it is only used to create the tokenizer. So can we assign it inside the factory by `model_type` instead of passing it in?

MorphlingEd replied:
The thing is the current `model_type` has only general information of the model but without the specific model name/path (e.g. `lm-sys/vicuna-7b-v1.5`). 
I might get what you mean by modifying `ModelType` to be a data class? Should we implement another class called like `ModelCard`, which has a field being `ModelType` in limited range and can have other information like model path?

dandansamax reviewed (changes requested):
This is an awesome change and we can invoke any model using openAI API after merging it!

There are some suggestions about code structure.

However, I think the biggest problem is that I should set up a entirely new environment to  launch a LLM server in my computer to use open source models, which is diffcult and time wasting. If I am a user who want to use a opensource model in camel, it is possibly because I don't have an openai key but I want to try camel for free. These complicated set up processes may prevent me from trying camel. We should either provide detailed instructions of how to set up a llm server or using a wrapper to do that in the code.

MorphlingEd commented on camel/models/open_source_model.py:
```diff
@@ -0,0 +1,102 @@
+from camel.typing import ModelType
+
+
+class OpenSourceModel(BaseModelBackend):
```
I agree with that this class has quite a lot in common with `OpenAIModel` but it introduces more checks currently. And if we need to do further modifications on the class for open-source models, it will be more convenient to have a separate class for them. I can currently modify it to inherit from the `OpenAIModel` to avoid some unnecessary copy-pastes while not making the two classes too complex.

dandansamax replied:
> And if we need to do further modifications on the class for open-source models, it will be more convenient to have a separate class for them.

The `BaseModelBackend` level is mainly related to APIs, so putting too much logic here may not be a good idea.

> I can currently modify it to inherit from the OpenAIModel to avoid some unnecessary copy-pastes while not making the two classes too complex.

This solution is better. In advance, how about creating a class for all OpenAI API models, including all open source models. Then let `OpenAIModel` inherit it, which is conceptually more clear than letting `OpenSourceModel` inherit.

dandansamax replied:
The main diffs are checking model_type, model_path, base_url. All of these checks can be removed if we correctly put these information in `model_config`.

lightaime replied:
Maybe we can find a better way to remove code duplication.

MorphlingEd commented on camel/typing.py:
```diff
@@ -29,10 +30,29 @@ class ModelType(Enum):
         return self.value if self.name != "STUB" else "gpt-3.5-turbo"
 
+    @property
+    def is_open_source(self) -> bool:
```
emm I think the aim to make this class `Enum` is to make the range of available model types restricted to what we currently support. If we change it to be a dataclass, will it be too flexible without limitations on the values?

dandansamax replied:
In my opinion, changing it to class will not remove limitation. For example, we have an abstract data class containing all information about a model (name, token_limit, base_url, open source or not, token counting function, etc). Then specific models inherit it and fill in these information. Users can either use our default model types by their classname or create a custom model by themselves.

dandansamax replied:
> It is an enumeration and we should keep it as an enumeration. It is not too much information. extending an enum with a property like is_open_source is fine.

Disagree. A enumeration containing too much information is not good.

MorphlingEd replied:
> I cannot agree with this idea. A enumeration containing too much information is not good.

I might get what you mean and write the code below. Could you have a look at the implementation below? This code isolate the data and behaviour of each model type into another dataclass.

e.g.
```python

@dataclass(frozen=True)
class ModelConfig:
    name: str = ...
    model_path: str = ...
    is_open_source: bool = ...
    ...

class ModelType(Enum):
    GPT_3_5_TURBO = "gpt-3.5-turbo"
    GPT_3_5_TURBO_16K = "gpt-3.5-turbo-16k"
    GPT_4 = "gpt-4"
    GPT_4_32k = "gpt-4-32k"
    STUB = "stub"

    LLAMA_2 = "llama-2"
    VICUNA = "vicuna"
    VICUNA_16K = "vicuna-16k"

class ModelConfigFactory:
    def create(model_type: ModelType, kwargs: Dict[str, Any]):
        if model_type in {
                ModelType.GPT_4,
                ModelType.GPT_3_5_TURBO,
                ...
        }:
            return ModelConfig(
                name = kwargs.get("model_name", model_type.value),
                model_path = None,
                is_open_source = False,
                ...
            )
       elif model_type in {
                ModelType.LLAMA_2,
                ...,
        }:
            if "model_path" not in kwargs:
                raise ValueError(...)

            return ModelConfig(
                name = kwargs.get("model_name", model_type.value),
                model_path = kwargs["model_path"]
                is_open_source = True,
                ...
            )
        else:
                raise ValueError("Invalid model type")
```

dandansamax replied:
Yeah. It's much more clear. Good enough for now.

MorphlingEd commented on camel/utils/token_counting.py:
```diff
@@ -0,0 +1,260 @@
+from camel.messages import OpenAIMessage
+from camel.typing import ModelType
+
+OPENAI_MODEL_VALUES = {
```
Will remove them. This is just used in a draft version😂

MorphlingEd commented on camel/utils/token_counting.py:
```diff
@@ -0,0 +1,260 @@
+        Returns:
+            BaseTokenCounter: The initialized token counter object.
+        """
+        if model_type in {
```
Yeah `is_open_source` is defined in `ModelType` and I will modify this part to directly use this function.

MorphlingEd commented:
> This is an awesome change and we can invoke any model using openAI API after merging it!
> 
> There are some suggestions about code structure.
> 
> However, I think the biggest problem is that I should set up a entirely new environment to launch a LLM server in my computer to use open source models, which is diffcult and time wasting. If I am a user who want to use a opensource model in camel, it is possibly because I don't have an openai key but I want to try camel for free. These complicated set up processes may prevent me from trying camel. We should either provide detailed instructions of how to set up a llm server or using a wrapper to do that in the code.
> 
> Anyway, this PR is very valuable for invoking other models, thank you so much!

Thanks for these valuable suggestions! Yeah I totally agree with the problem for the inconvenience launching the server, which I have discussed with Guohao. I am now trying to find a way to automatically launching a `FastChat` server given a model path. But after all, such an option will make users' options of the server limited to our choice.

Obs01ete commented on camel/models/open_source_model.py:
```diff
@@ -0,0 +1,102 @@
+        response = openai.ChatCompletion.create(messages=messages_openai,
+                                                model=self.model_name,
+                                                **self.model_config_dict)
+        if not self.stream:
```
Do opensource models support streaming mode? If not, please remove the code related to streaming.

Obs01ete replied:
Okay good to know, thanks

Obs01ete commented on camel/models/open_source_model.py:
```diff
@@ -0,0 +1,102 @@
+class OpenSourceModel(BaseModelBackend):
+    r"""OpenAI API in a unified BaseModelBackend interface."""
+
+    def __init__(self, model_type: ModelType, model_config_dict: Dict[str,
```
We need a test for this newly introduced class.

MorphlingEd commented on camel/agents/chat_agent.py:
```diff
@@ -113,6 +109,9 @@ class ChatAgent(BaseAgent):
         system_message (BaseMessage): The system message for the chat agent.
         model (ModelType, optional): The LLM model to use for generating
             responses. (default :obj:`ModelType.GPT_3_5_TURBO`)
+        model_path (str): The path to the folder containing the model file and
```
Good idea. I have thought about this but I thought as the `model_config` will be passed to OpenAI API as an argument, the `model_path` will cause errors. But I think I can just pop `model_path` during initializing `OpenSourceModel`

MorphlingEd commented:
Have just modified the current implementation but has not completed the documentation and tests. Will do in tomorrow morning as it is quite late (UTC+8)

lightaime reviewed (commented):
Awesome! Left some comments.

MorphlingEd commented on camel/typing.py:
```diff
@@ -49,9 +72,33 @@ def token_limit(self) -> int:
+            pattern = r'^vicuna-\d+b-v\d+\.\d+-16k$'
+            return bool(re.match(pattern, model_name))
+        else:
+            return self.value in model_name.lower()
```
This is because there are different versions and naming styles of the same model. For example, `Llama-2-7b-chat-hf` should have type `LLAMA_2` but its name is not exactly `llama-2`. The model name will be pased to the server for checking so we should not arbitrarily assign it.

Obs01ete replied:
A regex would be more appropriate here. 'in' can make mistakes.

MorphlingEd replied:
@Obs01ete But I am afraid that for LLaMA-2-based models, `in` might be sufficient for checking this match? I searched LLaMA-2-based models in HuggingFace and the names of them all contain "llama-2" or "llama2".

Obs01ete replied:
What if we have a local "camel-model-based-on-llama"?

lightaime replied:
Got it. I am good with this solution.

MorphlingEd commented on camel/societies/role_playing.py:
```diff
@@ -291,6 +291,8 @@ def init_agents(
                 function_call="auto",
             )
+        elif assistant_agent_kwargs:
+            assistant_config = assistant_agent_kwargs.pop("model_config", None)
```
Sorry that this was implemented in previous PR about function calling. This was because I thought the actual execution should only be done by the assistant while the user is focused on giving instructions without the need to do function calling. Does this make sense or do you have any suggestions on this?

MorphlingEd commented:
> > I cannot agree with this idea. A enumeration containing too much information is not good.
> 
> I might get what you mean and write the code below. Could you have a look at the implementation below? This code isolate the data and behaviour of each model type into another dataclass.
> 
> e.g.
> 
> ```python
> @dataclass(frozen=True)
> class ModelConfig:
>     name: str = ...
>     model_path: str = ...
>     is_open_source: bool = ...
>     ...
> 
> class ModelType(Enum):
>     GPT_3_5_TURBO = "gpt-3.5-turbo"
>     GPT_3_5_TURBO_16K = "gpt-3.5-turbo-16k"
>     GPT_4 = "gpt-4"
>     GPT_4_32k = "gpt-4-32k"
>     STUB = "stub"
> 
>     LLAMA_2 = "llama-2"
>     VICUNA = "vicuna"
>     VICUNA_16K = "vicuna-16k"
> 
> class ModelConfigFactory:
>     def create(model_type: ModelType, kwargs: Dict[str, Any]):
>         if model_type in {
>                 ModelType.GPT_4,
>                 ModelType.GPT_3_5_TURBO,
>                 ...
>         }:
>             return ModelConfig(
>                 name = kwargs.get("model_name", model_type.value),
>                 model_path = None,
>                 is_open_source = False,
>                 ...
>             )
>        elif model_type in {
>                 ModelType.LLAMA_2,
>                 ...,
>         }:
>             if "model_path" not in kwargs:
>                 raise ValueError(...)
> 
>             return ModelConfig(
>                 name = kwargs.get("model_name", model_type.value),
>                 model_path = kwargs["model_path"]
>                 is_open_source = True,
>                 ...
>             )
>         else:
>                 raise ValueError("Invalid model type")
> ```

Hi @Obs01ete @lightaime I write the sample code following Tianqi's suggestion and am wondering whether such an implementation of `ModelType` is viable? 
And some of the comments given by Guohao are mismatched with the commit done on yesterday (but I uploaded a commit after Guohao's comments without reading them today)

Obs01ete commented on camel/configs.py:
```diff
@@ -128,3 +128,8 @@ def from_openai_function_list(
+
+@dataclass(frozen=True)
+class OpenSourceConfig(ChatGPTConfig):
+    model_path: str = ""
```
Sure, we can port all dataclasses to pydantic in a separate commit. But: pydantic wont make any checks other than type checks afaik. One can just create a regular class with an explocit constructor for this.

Obs01ete commented:
@MorphlingEd Why do we need a factory for configs? Can't we make model backends in the factory and that's it?

MorphlingEd commented:
> @MorphlingEd Why do we need a factory for configs? Can't we make model backends in the factory and that's it?

Yeah that's a illed design shown here using a redundant factory. Later I tried to introduce a unified `ModelBackendConfig` which can be created with the model backend objects together in `ModelFactory`, containing all the information that is possessed by `ModelType` now. But I did not see the obvious convenience by completely removing properties from the `ModelType` enumeration. I need to write `ModelBackendConfig` for all of the already defined constants (e.g. `GPT_4`) one-by-one as they do not share properties. For example:
```python
...
if model_type is ModelType.GPT_4:
    return ModelConfig(
                ...
                token_limit = 8192,
                ...
            )
elif model_type is ModelType.GPT_3_5_TURBO:
    return ModelConfig(
                ...
                token_limit = 4096,
                ...
            )
```
Though I could use a dictionary like `token_limit_map` for mapping the model types to different token limits to reduce the repeated code, I think this is not simpler than using a "naive" `enum` class.

In addition, by introducing the new `OpenSourceConfig` in `config.py`, both the new arguments, `api_base` (`server_url`) and `model_path`, needed by open-source models can be conveniently passed. 
So I guess we can just keep the original implementation of `enum`?

MorphlingEd commented:
I just found there are some authentication problems due to the HuggingFace repo for LLaMA2 is gated. If we do not authenticate by the following commands in the check environment like below:
```shell
export HUGGINGFACE_TOKEN="<huggingface_token"
huggingface-cli login --token $HUGGINGFACE_TOKEN
```
When using HuggingFace's `AutoTokenizer`, it will report errors like:
`export HUGGINGFACE_TOKEN="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"
huggingface-cli login --token $HUGGINGFACE_TOKEN`

MorphlingEd commented:
Sorry just closed the branch by mistake

Obs01ete commented on camel/models/open_source_model.py:
```diff
@@ -0,0 +1,111 @@
+        """
+        super().__init__(model_type, model_config_dict)
+
+        import copy
```
All non-deferred imports to the top

Obs01ete commented on camel/models/open_source_model.py:
```diff
@@ -0,0 +1,111 @@
+        super().__init__(model_type, model_config_dict)
+
+        import copy
+        self.model_config_dict = copy.copy(model_config_dict)
```
Why do you need copy? deep copying stuff hints that you are doing something wrong

MorphlingEd replied:
Ahh.. this is because `model_config_dict` will be input as argument to OpenAI API and extra information like `model_path` and `server_url` will be poped from this dictionary. But I am afraid the change to the inpu dictionary object (or model config object) will result in some unexpected effect (e.g. the dict will be used later after initializing the agent). So I used a copy of the input argument.
Do you think whether this is necessary or not?

Obs01ete replied:
Please do not mutate variables. They are called variables but they must be as often constants as possible. Re-assemble the dictionary without unwanted keys with dict comprehension.

Obs01ete commented on camel/models/open_source_model.py:
```diff
@@ -0,0 +1,111 @@
+
+        # Check whether input model path is empty
+        model_path: str = self.model_config_dict.pop("model_path")
+        if model_path == "":
```
There should be no such things as empty strings. Either not present, or None, or a valid string with contents

Obs01ete commented on camel/models/open_source_model.py:
```diff
@@ -0,0 +1,111 @@
+        )
+
+        server_url = self.model_config_dict.pop("server_url")
+        if server_url == "":
```
There should be no such things as empty strings. Either not present, or None, or a valid string with contents

Obs01ete commented on camel/models/openai_model.py:
```diff
@@ -17,13 +17,19 @@
 
-    def __init__(self, model_type: ModelType,
-                 model_config_dict: Dict[str, Any]) -> None:
+    def __init__(
```
Revert

Obs01ete commented on camel/models/stub_model.py:
```diff
@@ -15,17 +15,21 @@
 
-    def __init__(self, model_type: ModelType,
-                 model_config_dict: Dict[str, Any]) -> None:
+    def __init__(
```
Revert

Obs01ete commented on camel/societies/role_playing.py:
```diff
@@ -286,13 +286,20 @@ def init_agents(
             output_language (str, optional): The language to be output by the
                 agents.
         """
+        import copy
```
Copy is no good. Why copying if a config must be an immutable thing?

Obs01ete reviewed (changes requested):
Great tests in test_open_source_model.py. Left some comments.

MorphlingEd commented on camel/configs.py:
```diff
@@ -128,3 +128,9 @@ def from_openai_function_list(
+@dataclass(frozen=True)
+class OpenSourceConfig(ChatGPTConfig):
+    model_path: str = ""
+    server_url: str = ""
```
Oh thanks for pointing this out! I just forgot the fields of a dataclass can exist without default values! 😂

MorphlingEd commented on camel/models/base_model.py:
```diff
@@ -31,6 +36,7 @@ def __init__(self, model_type: ModelType,
         """
         self.model_type = model_type
         self.model_config_dict = model_config_dict
+        self.token_counter: BaseTokenCounter
```
From this I have a thinking that the `TokenCounterFacotry` can be removed as it is not that convenient to initialize the token counter for both `OpenAIModel` and `OpenSourceModel`  in a unified way, where one requires just an empty dictionary while another one requires a dictionary containing `model_path` from the `model_config_dict`. 
My current thinking is to use an abstract method called `initialize_token_counter` and this method can just create the token counter object corresponding to the model backend class using functions in `token_counting.py`

Obs01ete replied:
I suggest an abstract read-only property token_counter. Then implementations will have `self._token_counter: Optional[BaseTokenCounter]` and an initialize-on-read property

MorphlingEd replied:
emm.. Does this mean that the `_token_counter` is initially `None`, and when `token_counter` is needed while the `_token_counter` is `None`, the `_token_counter` will then be initialized, e.g. using the implemented abstract method `initialize_token_counter`?

MorphlingEd replied:
But in this way I think there is no need to initialize-on-read. We can initialize the token counter when initializing the model while keeping the token counter as a read-only property

MorphlingEd commented:
@Obs01ete About the Regex of LLaMA2, though it may be like a joke, I would say "camel-model-based-on-llama" shoud belong to the type of LLaMA(1)? While the model based on LLaMA2 should be named as "camel-model-based-on-llama-2" (or "camel-model-based-on-llama2")?

lightaime commented on examples/open_source_models/role_playing_with_llama2.py:
I guess we do not need two examples since they are almost the same. Could we reduce them to one and put the other example as comments?

lightaime commented on examples/open_source_models/role_playing_with_llama2.py:
```diff
@@ -0,0 +1,90 @@
+from camel.utils import print_text_animated
+
+
+def main(model_type=None) -> None:
```
Let's put `model_path` and `model_path`here. And reduce the examples to one
```suggestion
def main(model_type=None, model_path=" ", serverl_url=" ") -> None:
```

lightaime commented on camel/utils/token_counting.py:
```diff
@@ -0,0 +1,214 @@
+        except:
+            raise ValueError(
+                f"Invalid `model_path` ({model_path}) is provided. "
+                "Tokenizer loading failed")
```
```suggestion
                "Tokenizer loading failed.")
```

lightaime commented on test/models/test_open_source_model.py:
```diff
@@ -0,0 +1,115 @@
+    with pytest.raises(
+            ValueError, match=re.escape(
+                ("Invalid `model_path` (vicuna-7b-v1.5) is provided. "
+                 "Tokenizer loading failed"))):
```
```suggestion
                 "Tokenizer loading failed."))):
```

lightaime reviewed (approved):
Approved it. Awesome @MorphlingEd! Like it! Left some minor comments.

lightaime commented on camel/models/model_factory.py:
```diff
@@ -50,8 +55,10 @@ def create(model_type: ModelType,
             model_class = OpenAIModel
         elif model_type == ModelType.STUB:
             model_class = StubModel
+        elif model_type.is_open_source:
+            model_class = OpenSourceModel
```
Should we also add `model_type.is_openai` as well?

MorphlingEd replied:
Sure. I have added one in `typing.py`.

MorphlingEd commented on camel/models/open_source_model.py:
```diff
@@ -0,0 +1,119 @@
+
+        Args:
+            model_type (ModelType): Model for which a backend is created,
+                one of GPT_* series.
```
Sorry for this mistake. I just copied the documentations but forgot to modify this part.

MorphlingEd commented:
I guess an instruction for setting up a server should be added to `README.md` (or a separate `README` in `models` directory?)

lightaime commented on examples/open_source_models/role_playing_with_open_source_model.py:
```diff
@@ -87,4 +85,11 @@ def main(model_type=None) -> None:
+    # Here :obj:`model_type` can be any of the supported open-source
+    # model types and :obj:`model_path` should be set corresponding to
+    # model type. 
+    main(
+        model_type=ModelType.LLAMA_2,
+        model_path="meta-llama/Llama-2-7b-chat-hf",
+        server_url="http://localhost:8000/v1",
+    )
```
Great! Can we also put vicuna example params here but comment out? So people can try it out as well.

MorphlingEd replied:
Will do :)

Obs01ete commented on camel/configs.py:
```diff
@@ -128,3 +128,16 @@ def from_openai_function_list(
+    server_url: str
+
+
+OPENAI_API_PARAMS = {param for param in asdict(ChatGPTConfig()).keys()}
```
In general I do not like computations in the global scope. But let it hang out there for this PR.

MorphlingEd replied:
I guess we can directly make this a set of string literals? Because the set of all arguments to OpenAI API is fixed.

Obs01ete commented on camel/configs.py:
```diff
@@ -128,3 +128,16 @@ def from_openai_function_list(
+
+
+@dataclass(frozen=True, kw_only=True)
+class OpenSourceConfig(ChatGPTConfig):
```
Please add docstrings

Obs01ete commented on camel/models/base_model.py:
```diff
@@ -30,25 +32,75 @@ def __init__(self, model_type: ModelType,
+        pass
+
+    @abstractmethod
+    def initialize_token_counter(self) -> BaseTokenCounter:
```
This method looks weird in the abstract class for a model backend. Can we remove it?

MorphlingEd replied:
emm...🤔Why is this weird? With this method I can just define the `token_counter` property in `BaseModel` without further modification in child classes. The child classes can focus on implementing how they initialize the token counter.

Obs01ete replied:
BaseModelBackend is an abstract class that specifies a so called contract between a user and an implementation. Some helper methods should not be there in the contract and are implementation-specific. Property token_counter is fine and is a good new interface method. But it must be it. Without strings attached like initialize_blah_blah.

Obs01ete commented on camel/models/open_source_model.py:
```diff
@@ -0,0 +1,140 @@
+                f"Model `{model_type}` is not a supported open-source model")
+
+        # Check whether input model path is empty
+        model_path: str | None = self.model_config_dict.get("model_path", None)
```
We do not support this style of type hints. Please rework for python 3.8 compatibility.

Obs01ete commented on controller.log.2023-08-10:
```diff
@@ -0,0 +1,464 @@
+2023-08-10 05:33:20 | INFO | controller | args: Namespace(host='localhost', port=21001, dispatch_method='shortest_queue')
```
Lol. Remove plz.

MorphlingEd replied:
Sorry about that. This was added by mistake 😂

Obs01ete reviewed (changes requested):
Overall good. Needs some extra polishing.

Obs01ete commented on camel/models/base_model.py:
```diff
@@ -30,25 +32,75 @@ def __init__(self, model_type: ModelType,
         self.model_config_dict = model_config_dict
+        self.check_model_config()
+
+        self._token_counter: Optional[BaseTokenCounter] = None
```
Move this to inheritants

MorphlingEd replied:
@Obs01ete Yeah I followed your suggestion and moved the definition `self._token_counter: Optional[BaseTokenCounter] = None` into the inheritants. But it seems each of the inheritant will repeat the same statement, does this matter? 🤔

Obs01ete replied:
It is fine in this case

Obs01ete commented on camel/models/base_model.py:
```diff
@@ -30,25 +32,75 @@ def __init__(self, model_type: ModelType,
+            BaseTokenCounter: The token counter following the model's
+                tokenization style.
+        """
+        if not self._token_counter:
```
Move implementation to inheritants

Obs01ete reviewed (changes requested):
Left my comments

lightaime commented on camel/societies/role_playing.py:
```diff
@@ -287,12 +287,16 @@ def init_agents(
-                function_list=self.assistant_functions,
-                function_call="auto",
-            )
+            assistant_model_config = (
+                FunctionCallingConfig.from_openai_function_list(
+                    function_list=self.assistant_functions,
+                    function_call="auto",
+                ))
+        elif assistant_agent_kwargs:
+            assistant_model_config = assistant_agent_kwargs.pop(
+                "model_config", None)
         else:
-            assistant_config = None
+            assistant_model_config = None
```
We might want to remove all these related to `configs` and `assistant_functions` https://github.com/camel-ai/camel/pull/249 since they can be passed from `*_agent_kwargs`.

lightaime commented on camel/societies/role_playing.py:
```diff
@@ -304,7 +308,7 @@ def init_agents(
         self.assistant_agent = ChatAgent(
             init_assistant_sys_msg,
-            model_config=assistant_config,
+            model_config=assistant_model_config,
```
We might want to remove all these related to `configs` and `assistant_functions` https://github.com/camel-ai/camel/pull/249 since they can be passed from `*_agent_kwargs`.

Obs01ete commented on camel/models/stub_model.py:
```diff
@@ -26,6 +26,19 @@ def __init__(self, model_type: ModelType,
+                tokenization style.
+        """
+        if not self._token_counter:
+            self._token_counter = OpenAITokenCounter(self.model_type)
```
Please create `StubTokenCounter(BaseTokenCounter)` class in this file that returns some constant (say int 10) from count_tokens_from_messages. We want to keep stubs super lightweight.

Obs01ete commented on camel/models/openai_model.py:
```diff
@@ -37,6 +37,19 @@ def __init__(self, model_type: ModelType,
                 be fed into openai.ChatCompletion.create().
         """
         super().__init__(model_type, model_config_dict)
+        self._token_counter: Optional[BaseTokenCounter] = None
```
Very good with this way of lazy initialization

Obs01ete commented on .github/workflows/documentation.yaml:
```diff
@@ -14,10 +14,10 @@ jobs:
 
     steps:
     - uses: actions/checkout@v3
-    - name: Set up Python 3.8
```
Please segregate the changes to GitHub actions into a separate MR. These changes are out of scope of this MR.

MorphlingEd replied:
🤔 But I did change the python version to 3.10 in `pyproject.toml` because the `kwonly` feature of dataclass is only available to versions after 3.10. Does this matter? (Without `kwonly=True` the `OpenSourceConfig` cannot have non-default values like `model_path`)

Obs01ete replied:
Definitely using a 3.10 feature of kwonly cannot be the reason to so easily bump the python version for the entire project. :D Can you rework the code to be 3.8 compatible? I am fine if you make a follow up discussion about why you need kwonly so badly that you want a bump to 3.10, then we will discuss it.

MorphlingEd commented:
@Obs01ete I have reverted the python version back to 3.8. But I failed trying to implement an equivalent functionality to `kw_only` even with a decorator. Finally I changed the `OpenSourceConfig` to be composed of an instance of `ChatGPTConfig`. This might be meaningful as configurations for open-source models should be different from `ChatGPTConfig` and the parameter values in `ChatGPTConfig` should be compacted into "parameters for OpenAI API". I am not sure whether this makes sense to you.

Obs01ete reviewed (approved):
I see, now all comments are resolved, and the code looks nice. Approved.

dandansamax commented:
LGTM. Only the config part is a little redundant now. We can consider refactoring it later. Anyway, good job! Feel free to merge it.

lightaime commented:
Could you wait for my final check a bit?

On Thu 24. Aug 2023 at 14:08, Dmitrii Khizbullin ***@***.***>
wrote:

> I changed the OpenSourceConfig to be composed of an instance of
> ChatGPTConfig
> Yeah, this does not look natural... Okay, let's leave it like this and
> rework later. Go ahead with merging this PR.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/camel-ai/camel/pull/245#issuecomment-1691556098>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AFUJTYHKLVWV6CYEHSIGHXDXW475VANCNFSM6AAAAAA3MPOFIA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

MorphlingEd commented:
> Could you wait for my final check a bit?


Yes of course :D

MorphlingEd commented:
@lightaime Hi Guohao, could I ask whether there is still something to be improved further? :)

lightaime commented on camel/models/open_source_model.py:
```diff
@@ -0,0 +1,143 @@
+                "URL to server running open-source LLM is not provided.")
+        self.server_url: str = server_url
+
+        # Replace :obj:model_config_dict with only the params to be
```
```suggestion
        # Replace `model_config_dict` with only the params to be
```

lightaime commented on camel/models/open_source_model.py:
```diff
@@ -0,0 +1,143 @@
+                "URL to server running open-source LLM is not provided.")
+        self.server_url: str = server_url
+
+        # Replace `model_config_dict`` with only the params to be
```
```suggestion
        # Replace `model_config_dict` with only the params to be
```

ShukriChiu commented:
@CodiumAI-Agent /review

CodiumAI-Agent commented:
## PR Analysis

- 🎯 **Main theme:** Integration of open-source models into the CAMEL framework
- 📝 **PR summary:** This PR introduces open-source models into the CAMEL framework, specifically LLaMA2 and Vicuna. The implementation assumes the user is running a server by himself and should provide CAMEL with the URL to this server. The PR also includes refactoring of the token counting functionality and removal of `BaseMessage.token_len` function.
- 📌 **Type of PR:** Enhancement
- 🧪 **Relevant tests added:** Yes
- 🔒 **Security concerns:** No security concerns found
## PR Feedback

- 💡 **General suggestions:** The PR is well-structured and includes a good amount of detail in the description. The integration of open-source models into the CAMEL framework is a significant enhancement. The refactoring of the token counting functionality and the removal of the `BaseMessage.token_len` function seem to be well-justified. However, it would be beneficial to include more comments in the code to explain the logic and the functionality of the methods, especially for the token counting functionality.


- **<details><summary> 🤖 Code feedback:**</summary>


  - **relevant file:** ``camel/utils/token_counting.py``
   **suggestion:** Consider adding more comments to explain the logic and functionality of the methods, especially for the token counting functionality. This will make the code more maintainable and understandable for other developers. [medium]
   **relevant line:** [class BaseTokenCounter(ABC):](https://github.com/camel-ai/camel/pull/245/files#diff-31358d0a5f05ccb1ca6db509b110493c9a82b7b36250ea447ebfdb41e6da9892R90)


  - **relevant file:** ``camel/utils/token_counting.py``
   **suggestion:** It would be beneficial to handle exceptions more gracefully in the `get_model_encoding` function. Instead of just printing "Model not found. Using cl100k_base encoding.", consider logging this information and also adding more context to the error message. [medium]
   **relevant line:** [def get_model_encoding(value_for_tiktoken: str):](https://github.com/camel-ai/camel/pull/245/files#diff-31358d0a5f05ccb1ca6db509b110493c9a82b7b36250ea447ebfdb41e6da9892R72)


  - **relevant file:** ``camel/utils/functions.py``
   **suggestion:** It seems like the `openai_api_key_required` function could be simplified. The function currently checks if the model is a stub or an open-source model, and if the OpenAI API key is in the environment variables. Consider simplifying this function to only check if the OpenAI API key is in the environment variables, as this is the only requirement for the function to work correctly. [medium]
   **relevant line:** [def openai_api_key_required(func: F) -> F:](https://github.com/camel-ai/camel/pull/245/files#diff-61b3061335d0b537a4158472a55dd411aa76b6da222472c7236a8d95a14e7273R39)


  - **relevant file:** ``camel/configs.py``
   **suggestion:** The `OpenSourceConfig` class could benefit from more detailed docstrings. Specifically, it would be helpful to explain what the `model_path` and `server_url` parameters are used for, and what their expected values are. [medium]
   **relevant line:** [class OpenSourceConfig(BaseConfig):](https://github.com/camel-ai/camel/pull/245/files#diff-c2b07a9e897382f9210ac6d674d8be0cd97f33025129103083b4eb80d6540e09R138)

</details>


### How to use
> Tag me in a comment '@CodiumAI-Agent' and add one of the following commands:
> **/review [-i]**: Request a review of your Pull Request. For an incremental review, which only considers changes since the last review, include the '-i' option.
> **/describe**: Modify the PR title and description based on the contents of the PR.
> **/improve [--extended]**: Suggest improvements to the code in the PR. Extended mode employs several calls, and provides a more thorough feedback. 
> **/ask \<QUESTION\>**: Pose a question about the PR.
> **/update_changelog**: Update the changelog based on the PR's contents.

>To edit any configuration parameter from **configuration.toml**, add --config_path=new_value
>For example: /review --pr_reviewer.extra_instructions="focus on the file: ..." 
>To list the possible configuration parameters, use the **/config** command.

maoyisier commented:
@CodiumAI-Agent /review

maoyisier commented:
@CodiumAI-Agent/review

CodiumAI-Agent commented:
## PR Analysis

- 🎯 **Main theme:** Integration of Open-source Models into CAMEL Framework
- 📝 **PR summary:** This PR introduces open-source models into the CAMEL framework. It includes the implementation of token counting for different types of open-source models and the necessary arguments for their specification. The PR also includes tests for the new functionality and refactors some existing code.
- 📌 **Type of PR:** Enhancement
- 🧪 **Relevant tests added:** Yes
- 🔒 **Security concerns:** No security concerns found
## PR Feedback

- 💡 **General suggestions:** The PR is well-structured and includes a significant amount of work. It is good to see that tests have been added for the new functionality. However, it would be beneficial to add more comments in the code to explain the logic and the purpose of the functions, especially for complex ones. This will make the code easier to understand and maintain in the future.


- **<details><summary> 🤖 Code feedback:**</summary>


  - **relevant file:** ``camel/utils/token_counting.py``
   **suggestion:** Consider handling exceptions more gracefully in the `OpenSourceTokenCounter` class. Instead of raising a ValueError when the tokenizer fails to load, you could log an error message and return a default value or handle the exception in a way that doesn't interrupt the execution of the program. [important]
   **relevant line:** ['+            raise ValueError(](https://github.com/camel-ai/camel/pull/245/files#diff-31358d0a5f05ccb1ca6db509b110493c9a82b7b36250ea447ebfdb41e6da9892R134)


  - **relevant file:** ``camel/utils/token_counting.py``
   **suggestion:** The `messages_to_prompt` function seems to have a lot of repeated code for different model types. Consider refactoring this function to reduce code duplication. You could create a dictionary to map model types to their corresponding separators and role maps, and then use this dictionary in the function. [medium]
   **relevant line:** ['+def messages_to_prompt(messages: List[OpenAIMessage], model: ModelType) -> str:'](https://github.com/camel-ai/camel/pull/245/files#diff-31358d0a5f05ccb1ca6db509b110493c9a82b7b36250ea447ebfdb41e6da9892R21)


  - **relevant file:** ``camel/utils/functions.py``
   **suggestion:** The `openai_api_key_required` decorator could be improved by using a more descriptive error message. Instead of just saying "OpenAI API key not found", you could provide more information about how to set the API key. [medium]
   **relevant line:** '+            raise ValueError("OpenAI API key not found")'


  - **relevant file:** ``camel/configs.py``
   **suggestion:** In the `OpenSourceConfig` class, it would be helpful to add type hints for the `model_path` and `server_url` parameters. This will make the code more readable and will allow static type checkers and IDE features to work correctly. [medium]
   **relevant line:** ['+    model_path: str'](https://github.com/camel-ai/camel/pull/245/files#diff-c2b07a9e897382f9210ac6d674d8be0cd97f33025129103083b4eb80d6540e09R150)

</details>


### How to use
> Tag me in a comment '@CodiumAI-Agent' and add one of the following commands:
> **/review [-i]**: Request a review of your Pull Request. For an incremental review, which only considers changes since the last review, include the '-i' option.
> **/describe**: Modify the PR title and description based on the contents of the PR.
> **/improve [--extended]**: Suggest improvements to the code in the PR. Extended mode employs several calls, and provides a more thorough feedback. 
> **/ask \<QUESTION\>**: Pose a question about the PR.
> **/update_changelog**: Update the changelog based on the PR's contents.

>To edit any configuration parameter from **configuration.toml**, add --config_path=new_value
>For example: /review --pr_reviewer.extra_instructions="focus on the file: ..." 
>To list the possible configuration parameters, use the **/config** command.

maoyisier commented:
@CodiumAI-Agent /describe

maoyisier commented:
@CodiumAI-Agent/describe

maoyisier commented:
/describe
